netket/models/slater.py

```diff
@@ -0,0 +1,112 @@
+# Copyright 2023 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import flax.linen as nn
+import jax.numpy as jnp
+
+from functools import partial
+
+from netket.experimental.hilbert import SpinOrbitalFermions
+from netket.utils.types import NNInitFunc, DType
+from netket.nn.masked_linear import default_kernel_init
+from netket import jax as nkjax
+
+
+def _log_det(A):
+    """Complex-aware logdeterminant built on top of
+    {func}`jax.numpy.linalg.slogdet` and combining the sign again.
+    """
+    sign, logabsdet = jnp.linalg.slogdet(A)
+    cplx_type = nkjax.dtype_complex(A.dtype)
+    return logabsdet.astype(cplx_type) + jnp.log(sign.astype(cplx_type))
+
+
+class LogSlater2nd(nn.Module):
```

Comment:
I would explicitly call this MeanFieldSlater2nd, so we can extend this with many body effects later. I would also drop the "Log" since all models are defined in log space.

Reply:
Indeed we can drop Log for consistency, though I wouldn't call it MeanField, I think Slater determinant is already widely understood to be mean field ? I don't like too much the `2nd` prefix we have a bit everywhere though, maybe `Discrete` seems a better option (also to align to DiscreteOperator etc?)

Reply:
Well, personally I'm a fan to keep the "2nd", since the idea was always that there would be the possibility to implement 1st quantization as well. The reason I'd add MeanField is simply that we will add other models as well probably that will use a general "Slater" module to compute slater determinants of backflowed orbitals..

Reply:
Right, thanks

Reply:
Indeed, that's why `Slater` should be the bare Slater determinant... Slater determinant is a synonym of a product state so I don't see the need to the qualifier MeanField (and SlaterProductState is even worse!) For other states we can call them `SlaterBackflow` etc...  There is no confusion there. 

Concerning the `2nd` I really find it inelegant too, but ok... I see the point. In my opinion, the solution is to have a general class without qualifiers `1st` or `2nd` etc that dispactes internally to the correct version, depending on the Hilbert. But we can do this addition later on